op-chain-ops/genesis/db_migration.go

```diff
@@ -52,98 +58,134 @@ func MigrateDB(ldb ethdb.Database, config *DeployConfig, l1Block *types.Block, m
+	}
+
+	// Convert all input messages into legacy messages. Note that this list is not yet filtered and
+	// may be missing some messages or have some extra messages.
 	unfilteredWithdrawals, err := migrationData.ToWithdrawals()
 	if err != nil {
 		return nil, fmt.Errorf("cannot serialize withdrawals: %w", err)
 	}
 
+	// We now need to check that we have all of the withdrawals that we expect to have. An error
+	// will be thrown if there are any missing messages, and any extra messages will be removed.
 	var filteredWithdrawals []*crossdomain.LegacyWithdrawal
 	if !noCheck {
 		log.Info("Checking withdrawals...")
-		filteredWithdrawals, err = PreCheckWithdrawals(db, unfilteredWithdrawals)
+		filteredWithdrawals, err = crossdomain.PreCheckWithdrawals(db, unfilteredWithdrawals)
 		if err != nil {
 			return nil, fmt.Errorf("withdrawals mismatch: %w", err)
 		}
-		log.Info("Withdrawals accounted for!")
 	} else {
 		log.Info("Skipping checking withdrawals")
 		filteredWithdrawals = unfilteredWithdrawals
 	}
 
-	// Now start the migration
-	log.Info("Setting the Proxies")
-	if err := SetL2Proxies(db); err != nil {
-		return nil, fmt.Errorf("cannot set L2Proxies: %w", err)
-	}
-
-	storage, err := NewL2StorageConfig(config, l1Block)
+	// We also need to verify that we have all of the storage slots for the LegacyERC20ETH contract
+	// that we expect to have. An error will be thrown if there are any missing storage slots.
+	// Unlike with withdrawals, we do not need to filter out extra addresses because their balances
+	// would necessarily be zero and therefore not affect the migration.
+	log.Info("Checking addresses...", "no-check", noCheck)
+	addrs, err := ether.PreCheckBalances(ldb, db, migrationData.Addresses(), migrationData.OvmAllowances, int(config.L1ChainID), noCheck)
 	if err != nil {
-		return nil, fmt.Errorf("cannot create storage config: %w", err)
+		return nil, fmt.Errorf("addresses mismatch: %w", err)
 	}
 
-	immutable, err := NewL2ImmutableConfig(config, l1Block)
-	if err != nil {
-		return nil, fmt.Errorf("cannot create immutable config: %w", err)
+	// At this point we've fully verified the witness data for the migration, so we can begin the
+	// actual migration process. This involves modifying parts of the legacy database and inserting
+	// a transition block.
+
+	// First order of business is to convert all predeployed smart contracts into proxies so they
+	// can be easily upgraded later on. In the legacy system, all upgrades to predeployed contracts
+	// required hard forks which was a huge pain. Note that we do NOT put the GovernanceToken or
+	// WETH9 contracts behind proxies because we do not want to make these easily upgradable.
+	log.Info("Converting predeployed contracts to proxies")
+	if err := SetL2Proxies(db); err != nil {
+		return nil, fmt.Errorf("cannot set L2Proxies: %w", err)
 	}
 
+	// Next we wipe the storage of all predeployed contracts EXCEPT for the GovernanceToken, WETH9,
```

Comment:
I don't think `SetImplementations` will wipe the storage explicitly. There's a check to see if the account exists, and if so it leaves the existing state alone:

```
if !db.Exist(codeAddr) {
    db.CreateAccount(codeAddr)
}
```

Reply:
This call to `db.CreateAccount` was the call the caused the earlier issue with state being erased.

Reply:
Ah yeah sorry this was written under the assumption that the other PR that mark was working on would make it in